library/common/http/dispatcher.cc

```diff
@@ -0,0 +1,108 @@
+#include "library/common/http/header_utility.h"
+
+namespace Envoy {
+namespace Http {
+
+Dispatcher::DirectStreamCallbacks::DirectStreamCallbacks(envoy_stream_t stream,
+                                                         envoy_observer observer,
+                                                         Dispatcher& http_dispatcher)
+    : stream_(stream), observer_(observer), http_dispatcher_(http_dispatcher) {}
+
+void Dispatcher::DirectStreamCallbacks::onHeaders(HeaderMapPtr&& headers, bool end_stream) {
+  ENVOY_LOG(debug, "response headers for stream (end_stream={}):\n{}", end_stream, *headers);
+  if (end_stream) {
+    http_dispatcher_.removeStream(stream_);
+  }
+  observer_.on_headers_f(Utility::transformHeaders(*headers), end_stream, observer_.context);
+}
+
+void Dispatcher::DirectStreamCallbacks::onData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(debug, "response data for stream (length={} end_stream={})", data.length(), end_stream);
+  if (end_stream) {
+    http_dispatcher_.removeStream(stream_);
+  }
+  observer_.on_data_f(Envoy::Buffer::Utility::transformData(data), end_stream, observer_.context);
+}
+
+void Dispatcher::DirectStreamCallbacks::onTrailers(HeaderMapPtr&& trailers) {
+  ENVOY_LOG(debug, "response trailers for stream:\n{}", *trailers);
+  http_dispatcher_.removeStream(stream_);
+  observer_.on_trailers_f(Utility::transformHeaders(*trailers), observer_.context);
+}
+
+void Dispatcher::DirectStreamCallbacks::onReset() {
+  http_dispatcher_.removeStream(stream_);
+  observer_.on_error_f({ENVOY_STREAM_RESET, {0, nullptr}}, observer_.context);
+}
+
+Dispatcher::DirectStream::DirectStream(AsyncClient::Stream& underlying_stream,
+                                       DirectStreamCallbacksPtr&& callbacks)
+    : underlying_stream_(underlying_stream), callbacks_(std::move(callbacks)) {}
+
+Dispatcher::Dispatcher(Event::Dispatcher& event_dispatcher,
+                       Upstream::ClusterManager& cluster_manager)
+    : current_stream_id_(0), event_dispatcher_(event_dispatcher),
+      cluster_manager_(cluster_manager) {}
+
+envoy_stream_t Dispatcher::startStream(envoy_observer observer) {
+  envoy_stream_t new_stream_id = current_stream_id_++;
+
+  event_dispatcher_.post([this, observer, new_stream_id]() -> void {
+    DirectStreamCallbacksPtr callbacks =
+        std::make_unique<DirectStreamCallbacks>(new_stream_id, observer, *this);
+    AsyncClient& async_client = cluster_manager_.httpAsyncClientForCluster("egress_cluster");
+    AsyncClient::Stream* underlying_stream = async_client.start(*callbacks, {});
+
+    if (!underlying_stream) {
+      callbacks->onReset();
+    } else {
+      DirectStreamPtr direct_stream =
+          std::make_unique<DirectStream>(*underlying_stream, std::move(callbacks));
+      streams_.emplace(new_stream_id, std::move(direct_stream));
+      ENVOY_LOG(debug, "started stream [{}]", new_stream_id);
+    }
+  });
+
+  return new_stream_id;
```

Comment:
Do you still want to return the stream ID in the case of an immediate failure above? Should the return value indicate the failure in some way? Either way can you document this case in the function docs?

Reply:
We can't synchronously know of immediate failure because it all runs on the dispatched post. The app will get the onReset() callback if failure to get a stream happens. We will document on the function docs.

Reply:
FYI it's possible this will happen before this function returns, so need to consider that as one of the cases in your cancellation follow ups.